### ISSUE TITLE
chatgpt failed to refresh auth token. 403 forbidden access

### ISSUE THREAD

**Issue by inuru** [barnesoir/chatgpt-vscode-plugin#22]
I have the token but always Got error 403. Is there step that I miss?

**Comment by zouweidong91** [comment]
+1

**Comment by iapYang** [comment]
+1

**Comment by minhlucvan** [comment]
+1

**Comment by mohammad-oghli** [comment]
+1

**Comment by mohammad-oghli** [comment]
If you check authentication session. It changed every second in the cookies tab. I think there is change in the backend code from the ChatGPT application side.

**Comment by minhlucvan** [comment]
refrence issue: https://github.com/transitive-bullshit/chatgpt-api/issues/96

**Comment by barnesoir** [comment]
Added a capture for cf_clearance token in latest release.

**Comment by ImYuya** [comment]
still getting the same error

**Comment by odragora** [comment]
Me too.
Followed the instructions in README, reset the token, pasted both tokens and still get 403: Forbidden.

**Comment by ImYuya** [comment]
@barnesoir
https://github.com/barnesoir/chatgpt-vscode-plugin/blob/d8ca1e44d94d8d3a531f4ccb952282b315e055c7/src/chatgpt-view-provider.ts#L61-L62

I'm not familiar with Typescript, but I think I need to fix it below.

```
this.sessionToken = userSessionToken!;
this.context.globalState.update('chatgpt-clearance-token', this.sessionToken);
```

↓

```
this.clearanceToken = userSessionToken!;
this.context.globalState.update('chatgpt-clearance-token', this.clearanceToken);
```

**Comment by timkmecl** [comment]
That's right, both the `clearanceToken` and `userAgent` have to be set and must match, as described [here](https://github.com/transitive-bullshit/chatgpt-api#update-december-12-2022). See [README.md](https://github.com/timkmecl/chatgpt-vscode#obtaining-the-three-tokens) from my extension for instructions on obtaining them.

**Comment by timkmecl** [comment]
My pull request #24 implements solutions by @mohammad-oghli and @ImYuya, should completely solve the issue caused by CloudFlare.

**Comment by ImYuya** [comment]
still getting the same error...
I input `sessionToken`, `clearanceToken` and `userAgent`.

**Comment by nilPogonin** [comment]
Same thing. Still same error

**Comment by mohammad-oghli** [comment]
+1 same issue.

**Comment by timkmecl** [comment]
Sorry, I've forgotten to commit changes that actually pass `userAgent` to `ChatGPTAPI` in the previous PR, #25 should fix this (I tested it and it works for me).

**Comment by minhlucvan** [comment]
![image](https://user-images.githubusercontent.com/11575076/207260466-f80f64a3-7720-42c0-808c-fc6dd0aa60b5.png)

**Comment by mohammad-oghli** [comment]
Not working for me.

**Comment by oldcodeoberyn** [comment]
the ChatGPTAPI update to 2.3.0, which require OpenAIEmail and OpenAIPassword to ensureAuth. **BUT!** it require nodejs >=18 which vscode bundle 16.14.3. 
No solution...

**Comment by timkmecl** [comment]
I just retested it and it still works for me. Maybe nodejs version is the problem like @oldcodeoberyn says; could someone with the problem post their node version (`node -v` in terminal) to check it? I have `v18.12.1`.

@mohammad-oghli Are you using it on the same machine and obtaining tokens in the same way as you did [here](https://github.com/timkmecl/chatgpt-vscode/pull/3#issuecomment-1346842416)? The implemented fix is basically the same as I did in my extension.

**Comment by timkmecl** [comment]
It looks like this problem is not limited to this project, but also to some other ones using chatgpt-api, probably caused by agressive CloudFlare protection. In my own testing, even when I tried to login to my account in Chrome right after using it inside Edge, I got a 403 error *inside the web version*. After I waited for several minutes and tried again, it disappeard, but this shows that the problem is not just with the javascript API.

The browser that you use for getting the tokens might influence the chance to get 403, especially chrome, so maybe try from another one (https://github.com/transitive-bullshit/chatgpt-api/issues/107#issuecomment-1349712906)

There might also be some account or IP flagging going on either by OpenAI or CloudFlare that increases the possibility of getting 403 especially if doing lots of requests often (https://github.com/transitive-bullshit/chatgpt-api/issues/96#issuecomment-1350275343)

### Update:

It seems like using Chrome to obtain the tokens is what is causing (most of?) these issues. So if you have 403 Error despite following the instructions, **try using a browser other than Chrome!**

**Comment by evilz** [comment]
It was not working with Chrome but it's ok using Edge (Win11) 👍🏻

**Comment by DavidRavnsborg** [comment]
I have tried in **Chrome, Firefox, Edge, and Brave**. They all lead to this error.

However, I haven't done JS development for a bit. My node version is `v14.16.1`. Going to try updating that and trying again.

**Comment by DavidRavnsborg** [comment]
I updated to node `v18.12.1` and am still encountering the same issue. Used the Brave browser.

E.g. useragent key-value pair:
`user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36`
